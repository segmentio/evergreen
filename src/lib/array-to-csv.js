import isEmpty from 'lodash.isempty'

/**
 * Converts an array to a comma separated string
 * @param {string[]} value
 * @returns {string}
 */
const arrayToCsv = value => {
  // Accept only array values + ensure there values inside the array, otherwise
  // there's nothing to join
  if (!Array.isArray(value) || isEmpty(value)) {
    return ''
  }

  return value.join(',')
}

export default arrayToCsv
