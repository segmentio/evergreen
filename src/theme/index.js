export { default as defaultTheme } from './src/default-theme'
export {
  default as ThemeContext,
  ThemeProvider,
  ThemeConsumer
} from './src/ThemeContext'
export { default as withTheme } from './src/withTheme'
export { default as useTheme } from './src/useTheme'
