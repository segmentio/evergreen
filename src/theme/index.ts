export { default as ThemeContext, getThemeContext, ThemeProvider, ThemeConsumer } from './src/ThemeContext'
export { default as withTheme } from './src/withTheme'
export { default as useTheme } from './src/useTheme'
export { get, getValue, mergeTheme, resolveThemeTokens } from './src/theme-tools'
