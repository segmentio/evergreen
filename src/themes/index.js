export { default as defaultTheme } from './default'
