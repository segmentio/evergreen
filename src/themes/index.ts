export { default as defaultTheme } from './default'
export type {
  Components,
  DefaultTheme,
  DefaultThemeAppearances,
  DefaultThemeColors,
  DefaultThemeColor,
  DefaultThemeFills,
  DefaultThemeIntents,
  DefaultThemeIntent,
  DefaultThemePseudoSelectors,
  DefaultThemeSizes,
} from './default'
export { default as classicTheme } from './classic'
export { default as deprecatedDefaultTheme } from './deprecated'
