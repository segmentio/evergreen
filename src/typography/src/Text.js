import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import FontFamilies from './styles/FontFamilies'
import TextStyles from './styles/TextStyles'
import TextUppercaseStyles from './styles/TextUppercaseStyles'
import TextColors from './styles/TextColors'

const isDev = process.env.NODE_ENV === 'development' || !process.env.NODE_ENV

export default class Text extends PureComponent {
  static propTypes = {
    ...Box.propTypes,
    size: PropTypes.oneOf(Object.keys(TextStyles).map(Number)),
    fontFamily: PropTypes.oneOf(Object.keys(FontFamilies)),
    textStyles: PropTypes.object,
    textUppercaseStyles: PropTypes.object,
    isUppercase: PropTypes.bool
  }

  static defaultProps = {
    size: 500,
    color: 'default',
    fontFamily: 'ui',
    textStyles: TextStyles,
    textUppercaseStyles: TextUppercaseStyles
  }

  render() {
    const {
      size,
      color,
      textStyles,
      textUppercaseStyles,
      fontFamily,
      isUppercase,
      ...props
    } = this.props

    let textStyle = textStyles[size]

    if (isUppercase) {
      // Only 100 and 200 support uppercase styles atm
      // Fallback on non uppercase atm
      if (Object.prototype.hasOwnProperty.call(textUppercaseStyles, size)) {
        textStyle = textUppercaseStyles[size]
      } else if (isDev) {
        console.error(
          `Uppercase ${size} not supported. <Text isUppercase> only supports the following sizes: ${JSON.stringify(
            Object.keys(textUppercaseStyles).map(Number)
          )}`
        )
      }
    }

    return (
      <Box
        is="span"
        {...(color ? { color: TextColors[color] || color } : {})}
        fontFamily={FontFamilies[fontFamily] || fontFamily}
        {...textStyle}
        {...props}
      />
    )
  }
}
