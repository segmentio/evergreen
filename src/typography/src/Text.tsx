import React, { forwardRef, memo } from 'react'
import Box, { PolymorphicBoxProps } from 'ui-box'
import { useStyleConfig } from '../../hooks'
import { useTheme } from '../../theme'
import { Size, FontFamily } from '../../types'

export type TextProps = PolymorphicBoxProps<'span', TextOwnProps>
export type TextOwnProps = {
  size?: Size
  fontFamily?: FontFamily | string
}

const emptyObject = {}

const Text: React.FC<TextProps> = memo(
  forwardRef(function Text(props, ref) {
    const { className, color: colorProp = 'default', fontFamily = 'ui', size = 400, ...restProps } = props

    const theme = useTheme()
    const { colors, fontFamilies } = theme

    const color = colorProp === 'none' || colorProp === 'default' ? 'default' : colorProp

    // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
    const themedFontFamily = fontFamilies[fontFamily] || fontFamily
    // @ts-expect-error ts-migrate(2538) FIXME: Type 'false' cannot be used as an index type.
    const themedColor = colors[color] || (colors.text && colors.text[color]) || color

    const textStyle = useStyleConfig('Text', { size }, emptyObject, emptyObject)

    return (
      <Box
        is="span"
        ref={ref}
        {...textStyle}
        fontFamily={themedFontFamily}
        color={themedColor}
        className={className}
        {...restProps}
      />
    )
  })
)

export default Text
