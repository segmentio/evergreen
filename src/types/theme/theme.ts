import { Components } from '../../themes'
import { Color } from './color'
import { ComponentStyles } from './component-styles'
import { Fill } from './fill'
import { FontFamilies } from './font-families'
import { FontWeights } from './font-weights'
import { Intent } from './intent'
import { LettingSpacings } from './letter-spacings'
import { ZIndices } from './z-indices'

export interface Theme<TComponents extends Components = Components> {
  colors: { [color: string]: Color<string | { [group: string]: Color }> }
  fills: { [fill: string]: Fill }
  intents: { [intent: string]: Intent }
  fontFamilies: FontFamilies
  radii: string[]
  shadows: string[] & { focusRing: string }
  fontSizes: string[]
  fontWeights: FontWeights
  letterSpacings: LettingSpacings
  lineHeights: string[]
  zIndices: ZIndices
  components: Partial<ComponentStyles<TComponents>>
}
