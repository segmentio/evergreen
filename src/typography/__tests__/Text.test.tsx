import React from 'react'
import { render } from '@testing-library/react'
// @ts-expect-error ts-migrate(7016) FIXME: Could not find a declaration file for module 'fake... Remove this comment to see the full error message
import faker from 'faker'
import renderer from 'react-test-renderer'
// @ts-expect-error ts-migrate(7016) FIXME: Could not find a declaration file for module '../.... Remove this comment to see the full error message
import { UIBoxSerializer } from '../../../lib/testing'
import { ThemeProvider } from '../../theme'
import { defaultTheme } from '../../themes'
import Text from '../src/Text'

// @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
expect.addSnapshotSerializer(UIBoxSerializer)

// @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'test'. Do you need to install ty... Remove this comment to see the full error message
test.each([
  ['size 300', 300],
  ['size 400', 400],
  ['size 500', 500],
  ['size 600', 600]
])('<Text /> %s renders as expected', (_: any, size: any) => {
  const component = (
    <ThemeProvider value={defaultTheme}>
      // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
      <Text size={size}>{`Text ${size}`}</Text>
    </ThemeProvider>
  )
  const tree = renderer.create(component).toJSON()
  // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
  expect(tree).toMatchSnapshot()
})

// @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'describe'. Do you need to instal... Remove this comment to see the full error message
describe('Colors', () => {
  // @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'test'. Do you need to install ty... Remove this comment to see the full error message
  test('<Text /> accepts arbitrary theme values for color', () => {
    const component = (
      <ThemeProvider value={defaultTheme}>
        // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <Text color="muted">Testing</Text>{' '}
      </ThemeProvider>
    )
    const tree = renderer.create(component).toJSON()
    // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
    expect(tree).toMatchSnapshot()
  })

  // @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'test'. Do you need to install ty... Remove this comment to see the full error message
  test('<Text /> does not render any color when a non-theme color is passed in ', () => {
    const component = (
      <ThemeProvider value={defaultTheme}>
        // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <Text color="SOMETHING DOESNT EXISt">Testing</Text>{' '}
      </ThemeProvider>
    )
    const tree = renderer.create(component).toJSON()
    // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
    expect(tree).toMatchSnapshot()
  })
})

// @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'describe'. Do you need to instal... Remove this comment to see the full error message
describe('Sizing', () => {
  const originalConsoleError = console.error
  // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'jest'.
  const mockFn = jest.fn()
  // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'beforeEach'.
  beforeEach(() => {
    console.error = mockFn
  })

  // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'afterEach'.
  afterEach(() => {
    console.error = originalConsoleError
  })

  // @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'test'. Do you need to install ty... Remove this comment to see the full error message
  test('<Text /> has undefined behavior when trying to set arbitrary sizes', () => {
    // @ts-expect-error ts-migrate(2322) FIXME: Type 'number' is not assignable to type 'never'.
    render(<Text size={800} />)
    // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
    expect(mockFn.mock.calls.length).toEqual(1)
    // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
    expect(mockFn.mock.calls[0][0]).toMatchInlineSnapshot(`
      Extracted Styles:
      box-sizing: border-box;
      color: #474d66;
      font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";


      "Warning: Failed %s type: %s%s"
    `)
  })
})

// @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'describe'. Do you need to instal... Remove this comment to see the full error message
describe('Props', () => {
  // @ts-expect-error ts-migrate(2582) FIXME: Cannot find name 'it'. Do you need to install type... Remove this comment to see the full error message
  it('should forward `className` prop', () => {
    const expected = faker.random.word().toLowerCase()
    const component = (
      // @ts-expect-error ts-migrate(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
      <Text data-testid="text" className={expected}>
        Testing
      </Text>
    )
    const { getByTestId } = render(component)
    // @ts-expect-error ts-migrate(2304) FIXME: Cannot find name 'expect'.
    expect(getByTestId('text')).toHaveClass(expected)
  })
})
