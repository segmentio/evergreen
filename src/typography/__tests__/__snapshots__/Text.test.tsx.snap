// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> size 300 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #474d66;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 12px;
font-weight: 400;
letter-spacing: 0;
line-height: 16px;


<span
  className="ub-fnt-sze_12px ub-f-wght_400 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-box-szg_border-box"
>
  Text 300
</span>
`;

exports[`<Text /> size 400 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #474d66;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 14px;
font-weight: 400;
letter-spacing: -0.05px;
line-height: 20px;


<span
  className="ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-fnt-fam_b77syt ub-color_474d66 ub-box-szg_border-box"
>
  Text 400
</span>
`;

exports[`<Text /> size 500 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #474d66;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
font-weight: 400;
letter-spacing: -0.05px;
line-height: 20px;


<span
  className="ub-fnt-sze_16px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-fnt-fam_b77syt ub-color_474d66 ub-box-szg_border-box"
>
  Text 500
</span>
`;

exports[`<Text /> size 600 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #474d66;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 18px;
font-weight: 400;
letter-spacing: -0.07px;
line-height: 24px;


<span
  className="ub-fnt-sze_18px ub-f-wght_400 ub-ln-ht_24px ub-ltr-spc_-0-07px ub-fnt-fam_b77syt ub-color_474d66 ub-box-szg_border-box"
>
  Text 600
</span>
`;

exports[`Colors <Text /> accepts arbitrary theme values for color 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #696f8c;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 14px;
font-weight: 400;
letter-spacing: -0.05px;
line-height: 20px;


Array [
  <span
    className="ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-fnt-fam_b77syt ub-color_696f8c ub-box-szg_border-box"
  >
    Testing
  </span>,
  " ",
]
`;

exports[`Colors <Text /> does not render any color when a non-theme color is passed in  1`] = `
Extracted Styles:
box-sizing: border-box;
color: SOMETHING DOESNT EXISt;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 14px;
font-weight: 400;
letter-spacing: -0.05px;
line-height: 20px;


Array [
  <span
    className="ub-fnt-sze_14px ub-f-wght_400 ub-ln-ht_20px ub-ltr-spc_-0-05px ub-fnt-fam_b77syt ub-color_SOMETHING-DOESNT-EXISt ub-box-szg_border-box"
  >
    Testing
  </span>,
  " ",
]
`;
